torchacc/utils/optim_utils.py

```diff
@@ -0,0 +1,225 @@
+import torch
+import torch.distributed as dist
+from typing import Any, Dict, NamedTuple, Optional
+from torch.utils._pytree import tree_map_only
+import torch_xla.core.xla_model as xm
+
+
+def _numel(shape):
+    numel = 1
+    for d in shape:
+        numel *= d
+    return numel
+
+
+def get_layer_full_info(shard_metadata, model_state_dict):
+    """
+    Get full name, shape and numel info of unflatten and unshard optimizer's state_dict according 
+    to shard_metadata and model's state_dict;
+    Args:
+        shard_metadata (dict):
+            ``model.get_shard_metadata()`` from an FSDP model of any rank
+        model_state_dict(dict):
+            The state_dict from an FSDP model.
+
+    Returns:
+        For all ranks, we get the same shard_metadata and model_state_dict, and the return value is
+        same:
+        layer_name_list(list): 2-dimension list([[layer_name_group1], [layer_name_group2], ...]), contains the full name information.
+        if parameters if flatten, each layer may have mutiple orig name and parameter.
+        layer_size_list(list): 2-dimension list([[layer_size_group1], [layer_size_group2], ...]), contains the unflatten and unshard shape information of 
+        each layer.
+        layer_numel_list(list): 2-dimension list([[layer_numel_group1], [layer_numel_group2], ...]), contains the unflatten and unshard numel information of 
+        each layer. 
+    """
+    layer_name_list = []
+    layer_size_list = []
+    layer_numel_list = []
+    buffer_info = shard_metadata.get("buffer_info", {})
+
+    # consolidate the sharded parameters
+    for name, param in model_state_dict.items():
+        if name in buffer_info:  # cast buffer back to its original dtype
+            p = p.to(buffer_info[name]["_orig_dtype"])
+
+        is_sharded = False
+        name_splits = name.split(".")
+        # if start with 'model', we just skip the 'model'
+        if name_splits[0] == 'model':
+            name = ".".join(name_splits[1:])
+            name_splits = name.split(".")
+        for idx, sep in enumerate(name_splits):
+            if sep.startswith("_fsdp_shard"):
+                is_sharded = True
+                prefix = ".".join(name_splits[:idx])
+                suffix = ".".join(name_splits[idx:])
+                break
+
+        if is_sharded:
+            p_info = shard_metadata["shard_info"][prefix][suffix]
+            orig_name = p_info["_orig_name"]
+            orig_size = p_info["_orig_size"]
+            full_name = orig_name
+            if prefix != "":
+                full_name = prefix + "." + orig_name
+            layer_name_list.append(full_name)
+            layer_size_list.append(orig_size)
+            layer_numel_list.append(_numel(orig_size))
+
+        else:
+            # unsharded buffers, we don't need the info in shard_metadata
```

Comment:
I refer to the implementation here: https://github.com/AlibabaPAI/xla/blob/a32543b7417d1b7b26a99e07d50ea4437d16323c/torch_xla/distributed/fsdp/state_dict_utils.py#L87. In FSDP, there may have unsharded buffers?